lsp-headerline.el

```diff
@@ -0,0 +1,345 @@
+                              ('path-up-to-project (or lsp--headerline-breadcrumb-path-up-to-project-string
+                                                       (lsp--headerline-breadcrumb-build-path-up-to-project-string)))
+                              ('symbols (lsp--headerline-breadcrumb-build-symbols-string)))
+                            "")))
+              lsp-headerline-breadcrumb-segments ""))
+
+(defun lsp--document-symbols->symbols-hierarchy (document-symbols)
+  "Convert DOCUMENT-SYMBOLS to symbols hierarchy."
+  (-let (((symbol &as &DocumentSymbol? :children?)
+          (seq-some (-lambda ((symbol &as &DocumentSymbol :range (&RangeToPoint :start :end)))
+                      (when (<= start (point) end)
+                        symbol))
+                    document-symbols)))
+    (if children?
+        (cons symbol (lsp--document-symbols->symbols-hierarchy children?))
+      (when symbol
+        (list symbol)))))
+
+(defun lsp--symbols-informations->symbols-hierarchy (symbols-informations)
+  "Convert SYMBOLS-INFORMATIONS to symbols hierarchy."
+  (seq-filter (-lambda ((symbol &as &SymbolInformation :location (&Location :range (&RangeToPoint :start :end))))
+                (when (<= start (point) end)
+                  symbol))
+              symbols-informations))
+
+(defun lsp-symbols->symbols-hierarchy (symbols)
+  "Convert SYMBOLS to symbols-hierarchy."
+  (when-let (first-symbol (lsp-seq-first symbols))
+    (if (lsp-symbol-information? first-symbol)
+        (lsp--symbols-informations->symbols-hierarchy symbols)
+      (lsp--document-symbols->symbols-hierarchy symbols))))
+
+(defun lsp-headerline--check-breadcrumb (&rest _)
+  "Request for document symbols to build the breadcrumb."
+  (setq lsp--headerline-breadcrumb-string (lsp--headerline-build-string))
+  (force-mode-line-update))
+
+(defun lsp-headerline--breadcrumb-cache-path-up-to-project ()
+  "Cache the path-up-to-project breadcrumb segment if enabled."
+  (when (and lsp-headerline-breadcrumb-enable
+             (member 'path-up-to-project lsp-headerline-breadcrumb-segments))
+    (setq lsp--headerline-breadcrumb-path-up-to-project-string (lsp--headerline-breadcrumb-build-path-up-to-project-string))))
+
+(defun lsp--headerline-enable-breadcrumb ()
+  "Enable headerline breadcrumb mode."
+  (when (and lsp-headerline-breadcrumb-enable
+             (lsp-feature? "textDocument/documentSymbol"))
+    (lsp-headerline-breadcrumb-mode 1)))
+
+(defun lsp--headerline-disable-breadcrumb ()
+  "Disable headerline breadcrumb mode."
+  (lsp-headerline-breadcrumb-mode -1))
+
+;;;###autoload
+(define-minor-mode lsp-headerline-breadcrumb-mode
+  "Toggle breadcrumb on headerline."
+  :group 'lsp-mode
+  :global nil
+  (cond
+   (lsp-headerline-breadcrumb-mode
+    (add-to-list 'header-line-format '(t (:eval lsp--headerline-breadcrumb-string)))
+    (add-hook 'lsp-managed-mode-hook #'lsp-headerline--breadcrumb-cache-path-up-to-project nil t)
```

Comment:
Shouldn't `lsp-headerline--breadcrumb-cache-path-up-to-project` be called per `lsp-headerline-breadcumb-mode` instead?

Reply:
Yes, I think so, I can change to run when this mode is enabled so

Reply:
rethinking, I'm not sure if we can just call `lsp-headerline--breadcrumb-cache-path-up-to-project` per this mode initialization as lsp can shutdown then reconnect and we may want to recalculate that, maybe depending on managed-hook is safer.

Reply:
When you shutdown lsp and reconnect, shouldn't this mode initialization run again?
That's the whole point of `lsp--configure-buffer` isn't it?

Reply:
Yeah, I think it's ok too, change to run when the mode is enabled